### ISSUE TITLE
[2019/12/19] 推薦・機械学習勉強会

### ISSUE THREAD

**Issue by yu-ya4** [wantedly/machine-learning-round-table#30]
## Why

この会は、推薦や機械学習、その周辺技術を通じてサービスを改善することにモチベーションのある人達の集まりです。

発信のため、ここは **public** にしてあります。外部からの参加をご希望の方は Wantedly よりご連絡ください！

## What

Wantedly では隔週木曜日に

- 推薦の評価指標について議論したい
- 〇〇っていうライブラリ / フレームワークを導入してみたい
- 他社の基盤事例をみんなにシェアして自社の基盤開発に活かしたい
- もっと推薦を良くするためにどんなものが必要か議論したい

といった話をする「推薦・機械学習勉強会」を開催しています。
この ISSUE はその会で話すネタを共有するための場所です。

話したいことがある人はここにコメントしましょう！
会の間に話した内容もここにメモしましょう！

**Comment by yu-ya4** [comment]
## Reciprocal Recommender Systems（相互推薦システム）の紹介

４兆年ぶりにブログ書いた。

- Reciprocal Recommender System の概要
- 従来の Recommneder Systems との性質比較
- システム構成
- これまでの提案手法紹介

http://yu-ya4.hatenablog.com/entry/2019/12/12/184120

**Comment by yuko-i** [comment]
# Kaggle Days Workshop - Practical tips for handling noisy data and annotation -  

このワークショップ（ノイジーなデータを処理するためのヒントをテーマにしたワークショップ）に関して[ブログ](https://www.wantedly.com/companies/wantedly/post_articles/199197)でも少し書いたのですが、 (長くなると思って)書ききれなかったことについて、共有します。

## Mixup
* manifold mixup (中間層でMixupする手法: [参考 Qiita](https://qiita.com/kirikei/items/1fb56f22f4f48c5f91f0) )は画像データに限ったことではなく、テーブルデータのカテゴリーデータを中間層で混ぜるとかも可能
* ミニバッチの中で、データをシャッフルすると実装しやすい
* 最後の方だけ、augumentationを止めると良いと言う論文が出たのでその通りやったら少し改善した

## Large batch size
そもそも、バッチサイズを大きくしたくても、GPUメモリに乗り切らなければ大きくできない問題がある。
以下の方法で対応。
* Gradient accumulation: 重みを毎ステップ更新するのではなく、数ステップ勾配情報を貯めておいて、重み更新をする。 (擬似的にバッチサイズを大きくさせる)ただし、BNはバッチごとのデータで行われるで、厳密には同一ではない。
* Mixed Precision Training: apexを利用。質疑応答の中でMixed Precision Trainingだと精度が劣るのでは？という質問があったが、発表者の方（Kanoh さん）の経験的にはMixed Precision Trainingを使ってバッチサイズを大きくした方が結果的に良かった。

MixupとDistillationがノイジーデータに有効であることをこのワークショップで知れたので良かった。

[プレゼン内容: SlideShare](https://www.slideshare.net/RyuichiKanoh/practical-tips-for-handling-noisy-data-and-annotaiton-204195412)

**Comment by taro-masuda** [comment]
## Kaggle Days Workshop - Feature engineering for events data -
### Recsys2019 1st place solution
- trivagoのホテル予約サイトのデータセット
- MRR (Mean Reciprocal Rank) という評価指標。こちらの記事が分かりやすい
  - http://blog.brainpad.co.jp/entry/2017/08/25/140000#MRRMean-Reciprocal-Rank
  - ユーザごとに最初に適合したアイテムの順位の逆数の平均を取るため、値域は[0,1]
- 過学習しにくいデータセットだったので、観測データを100万件に小さくして素早くValidationを実施したのち、上手くいったら全件でテストするという戦略をとった
- 特徴量はクリック数やCTRなどほとんどが集計量（accumulators）で、テスト時よりも過去のデータしか使えない点に注意
- CTRの相関も使う; 例えば25個のホテルのリストが表示されているとき、5番目のホテルをクリックしたら近辺の6,7,8番目のホテルのクリックが共起するかどうかを特徴量としている
- Leakyな特徴を見つけた; 各アイテムの位置情報に加えて、1つ前のクリックの位置とそのtimestampだけからMRR=0.63くらい予測できる
    - click location と timestamp があった
    - T_diff == 0 -> 同じ場所をdouble clickなどでクリックしたものが99%
    - T_diff <= 5s ->「次へ」ボタンのように1個先をクリックするアクションが大半
    - T_diff <= 2min -> 何個先をクリックしているかはユーザによりけりで、正規分布に近づく
- 最後2週間でモデルのチューニング
  - DARTに切り替え（学習は遅いが精度が良かったらしい）
    - DART: NNのドロップアウトの考えを取り入れたブースティング木
      - https://xgboost.readthedocs.io/en/latest/tutorials/dart.html
      - `import xgboost as xgb` だけで試せる
  - ハイパラチューニング（Learning rate, 木のdepth, drop rate）

**Comment by taro-masuda** [comment]
> # Kaggle Days Workshop - Practical tips for handling noisy data and annotation -
> このワークショップ（ノイジーなデータを処理するためのヒントをテーマにしたワークショップ）に関して[ブログ](https://www.wantedly.com/companies/wantedly/post_articles/199197)でも少し書いたのですが、 (長くなると思って)書ききれなかったことについて、共有します。
> 
> ## Mixup
> * manifold mixup (中間層でMixupする手法: [参考 Qiita](https://qiita.com/kirikei/items/1fb56f22f4f48c5f91f0) )は画像データに限ったことではなく、テーブルデータのカテゴリーデータを中間層で混ぜるとかも可能
> * ミニバッチの中で、データをシャッフルすると実装しやすい
> * 最後の方だけ、augumentationを止めると良いと言う論文が出たのでその通りやったら少し改善した
> 
> ## Large batch size
> そもそも、バッチサイズを大きくしたくても、GPUメモリに乗り切らなければ大きくできない問題がある。
> 以下の方法で対応。
> 
> * Gradient accumulation: 重みを毎ステップ更新するのではなく、数ステップ勾配情報を貯めておいて、重み更新をする。 (擬似的にバッチサイズを大きくさせる)ただし、BNはバッチごとのデータで行われるで、厳密には同一ではない。
> * Mixed Precision Training: apexを利用。質疑応答の中でMixed Precision Trainingだと精度が劣るのでは？という質問があったが、発表者の方（Kanoh さん）の経験的にはMixed Precision Trainingを使ってバッチサイズを大きくした方が結果的に良かった。
> 
> MixupとDistillationがノイジーデータに有効であることをこのワークショップで知れたので良かった。
> 
> [プレゼン内容: SlideShare](https://www.slideshare.net/RyuichiKanoh/practical-tips-for-handling-noisy-data-and-annotaiton-204195412)

- Mixed Precision ->　例えば16bit(半精度)で高速に回したいときに一時的に16bitが細かく表現しやすいスケールに疑似的に移す手法だそう…

**Comment by taro-masuda** [comment]
## 新データ分析コンペプラットフォームNishika
- https://www.nishika.com/competitions/2/ranking

**Comment by yu-ya4** [comment]
## Research Methods in Machine Learning

http://web.engr.oregonstate.edu/~tgd/talks/new-in-ml-2019.pdf

- 研究テーマの決め方の話
  - Paper-Driven Revision of the Method
- 研究テーマのライフサイクルを意識して選択することが重要
  - 実験的課題設定
  - ナイーブな解決手法の提示
  - 手法の洗練/評価の提示
  - 改善競争
  - 課題領域の特定と解決策の対応付け
  - エンジニアリング(実応用)
- 各 Phase ごとに具体例が豊富
- それぞれ必要なスキルが別だよね

<img width="1180" alt="スクリーンショット 2019-12-19 18 38 35" src="https://user-images.githubusercontent.com/12006023/71162580-d0ae3680-228e-11ea-8810-5b37206c804b.png">

**Comment by taro-masuda** [comment]
https://twitter.com/nkmk_me/status/1067984125359845376?s=20
```
nkmknkmk @nkmk_me

#pandas豆知識
「http://pd.np」でpandasからNumPyの関数や値を使用できる。pdをインポートすればnpを明示的にインポートする必要はない。
Jupyter Notebookで書き捨てのコードを書いているとき、途中で一回だけNumPy関数を使いたくなった場合などに便利。

https://note.nkmk.me/python-pandas-pd-np/
```

**Comment by taro-masuda** [comment]
## MeCab-pythonのメンテナンスが終了
- 代替となるfugashiはipadic-neologdが使えない（e.g. 「銀河鉄道の夜」をどうパースするか問題提起をしているのでアンチテーゼ的か）

**Comment by taro-masuda** [comment]
## 楽曲推薦ネタ
- https://twitter.com/naotakafujii/status/1207131996993929216?s=21

**Comment by taro-masuda** [comment]
## 次世代Batch Normalization戦争
- Local Context Normalization
  - https://arxiv.org/abs/1912.05845

**Comment by yu-ya4** [comment]
## 人類が皆知っておくべき ランキング評価手法 Interleaving

- A/Bテストより10~100倍効率的なランキング評価手法　インターリービング（Interleaving）のまとめと実践
  - https://qiita.com/mpkato/items/99bd55cc17387844fd62
- 安心・高速・正確な2019年最新のインターリービング手法 Pairwise Preference Multileaving
  - https://qiita.com/mpkato/items/cc7fc36d74268af130d5
